In [2]:
import warnings
warnings.filterwarnings("ignore")

# Data Cleaning

## Install Requirements

In [ ]:
!pip install -r requirements.txt

---
## Import and Settings

In [16]:
import numpy as np

In [4]:
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [5]:
import dask
import dask.dataframe as dd
import dask.array as da

In [6]:
df = dd.read_csv('unbalaced_20_80_dataset.csv')

---
## Data Overview

In [7]:
df.head()

Unnamed: 0                                 Flow ID          Src IP  \
0     1739476   172.31.69.25-18.219.193.20-80-37882-6   18.219.193.20   
1     1822666      172.31.69.28-18.219.9.1-80-63287-6    172.31.69.28   
2      905739   172.31.69.28-52.14.136.135-80-63095-6   52.14.136.135   
3     1143064  172.31.69.28-18.216.200.189-80-52341-6  18.216.200.189   
4     1934016   172.31.69.28-18.218.55.126-80-57459-6    172.31.69.28   

   Src Port         Dst IP  Dst Port  Protocol               Timestamp  \
0     37882   172.31.69.25        80         6  16/02/2018 11:27:29 PM   
1        80     18.219.9.1     63287         6  22/02/2018 12:13:52 AM   
2     63095   172.31.69.28        80         6  22/02/2018 12:14:02 AM   
3     52341   172.31.69.28        80         6  22/02/2018 12:28:04 AM   
4        80  18.218.55.126     57459         6  22/02/2018 12:19:45 AM   

   Flow Duration  Tot Fwd Pkts  Tot Bwd Pkts  TotLen Fwd Pkts  \
0           8660             1             1              0.0   
1           5829             4             3            935.0   
2           3396             1             1              0.0   
3           2390             1             1              0.0   
4          17362             4             3            935.0   

   TotLen Bwd Pkts  Fwd Pkt Len Max  Fwd Pkt Len Min  Fwd Pkt Len Mean  \
0              0.0              0.0              0.0              0.00   
1            298.0            935.0              0.0            233.75   
2              0.0              0.0              0.0              0.00   
3              0.0              0.0              0.0              0.00   
4            314.0            935.0              0.0            233.75   

   Fwd Pkt Len Std  Bwd Pkt Len Max  Bwd Pkt Len Min  Bwd Pkt Len Mean  \
0              0.0              0.0              0.0          0.000000   
1            467.5            298.0              0.0         99.333333   
2              0.0              0.0              0.0          0.000000   
3              0.0              0.0              0.0          0.000000   
4            467.5            314.0              0.0        104.666667   

   Bwd Pkt Len Std    Flow Byts/s  Flow Pkts/s  Flow IAT Mean  Flow IAT Std  \
0         0.000000       0.000000   230.946882    8660.000000      0.000000   
1       172.050380  211528.564076  1200.892091     971.500000   2104.124973   
2         0.000000       0.000000   588.928151    3396.000000      0.000000   
3         0.000000       0.000000   836.820084    2390.000000      0.000000   
4       181.287985   71938.716738   403.179357    2893.666667   6770.577661   

   Flow IAT Max  Flow IAT Min  Fwd IAT Tot  Fwd IAT Mean  Fwd IAT Std  \
0        8660.0        8660.0          0.0      0.000000     0.000000   
1        5260.0           7.0       5822.0   1940.666667  3119.411857   
2        3396.0        3396.0          0.0      0.000000     0.000000   
3        2390.0        2390.0          0.0      0.000000     0.000000   
4       16711.0           7.0      17355.0   5785.000000  9779.680414   

   Fwd IAT Max  Fwd IAT Min  Bwd IAT Tot  Bwd IAT Mean   Bwd IAT Std  \
0          0.0          0.0          0.0           0.0      0.000000   
1       5541.0         46.0       5540.0        2770.0   3521.391770   
2          0.0          0.0          0.0           0.0      0.000000   
3          0.0          0.0          0.0           0.0      0.000000   
4      17077.0         38.0      17074.0        8537.0  11559.781659   

   Bwd IAT Max  Bwd IAT Min  Fwd PSH Flags  Bwd PSH Flags  Fwd URG Flags  \
0          0.0          0.0              0              0              0   
1       5260.0        280.0              0              0              0   
2          0.0          0.0              0              0              0   
3          0.0          0.0              0              0              0   
4      16711.0        363.0              0              0              0   

   Bwd URG Flags  Fwd Header Len  

In [8]:
df.columns

Index(['Unnamed: 0', 'Flow ID', 'Src IP', 'Src Port', 'Dst IP', 'Dst Port',
       'Protocol', 'Timestamp', 'Flow Duration', 'Tot Fwd Pkts',
       'Tot Bwd Pkts', 'TotLen Fwd Pkts', 'TotLen Bwd Pkts', 'Fwd Pkt Len Max',
       'Fwd Pkt Len Min', 'Fwd Pkt Len Mean', 'Fwd Pkt Len Std',
       'Bwd Pkt Len Max', 'Bwd Pkt Len Min', 'Bwd Pkt Len Mean',
       'Bwd Pkt Len Std', 'Flow Byts/s', 'Flow Pkts/s', 'Flow IAT Mean',
       'Flow IAT Std', 'Flow IAT Max', 'Flow IAT Min', 'Fwd IAT Tot',
       'Fwd IAT Mean', 'Fwd IAT Std', 'Fwd IAT Max', 'Fwd IAT Min',
       'Bwd IAT Tot', 'Bwd IAT Mean', 'Bwd IAT Std', 'Bwd IAT Max',
       'Bwd IAT Min', 'Fwd PSH Flags', 'Bwd PSH Flags', 'Fwd URG Flags',
       'Bwd URG Flags', 'Fwd Header Len', 'Bwd Header Len', 'Fwd Pkts/s',
       'Bwd Pkts/s', 'Pkt Len Min', 'Pkt Len Max', 'Pkt Len Mean',
       'Pkt Len Std', 'Pkt Len Var', 'FIN Flag Cnt', 'SYN Flag Cnt',
       'RST Flag Cnt', 'PSH Flag Cnt', 'ACK Flag Cnt', 'URG Flag Cnt',
       'CWE Flag

In [10]:
df.info(verbose=True)

<class 'dask.dataframe.core.DataFrame'>
Int64Index: 7616509 entries, 0 to 51350
Data columns (total 85 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   Unnamed: 0         7616509 non-null      int64
 1   Flow ID            7616509 non-null      object
 2   Src IP             7616509 non-null      object
 3   Src Port           7616509 non-null      int64
 4   Dst IP             7616509 non-null      object
 5   Dst Port           7616509 non-null      int64
 6   Protocol           7616509 non-null      int64
 7   Timestamp          7616509 non-null      object
 8   Flow Duration      7616509 non-null      int64
 9   Tot Fwd Pkts       7616509 non-null      int64
10   Tot Bwd Pkts       7616509 non-null      int64
11   TotLen Fwd Pkts    7616509 non-null      float64
12   TotLen Bwd Pkts    7616509 non-null      float64
13   Fwd Pkt Len Max    7616509 non-null      float64
14   Fwd Pkt Len Min    7616509 non-null      float64
15 

In [11]:
df.describe().compute()

Unnamed: 0      Src Port      Dst Port      Protocol  Flow Duration  \
count  7.616509e+06  7.616509e+06  7.616509e+06  7.616509e+06   7.616509e+06   
mean   3.518141e+06  4.171610e+04  1.030737e+04  9.066993e+00   1.168721e+07   
std    2.325547e+06  2.326642e+04  2.013310e+04  5.087024e+00   3.066261e+07   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  -1.000000e+00   
25%    1.985860e+06  4.947100e+04  8.000000e+01  6.000000e+00   2.915000e+03   
50%    3.954691e+06  5.213900e+04  4.430000e+02  6.000000e+00   3.736400e+04   
75%    5.941020e+06  5.755400e+04  4.973375e+04  1.700000e+01   4.289224e+06   
max    7.902474e+06  6.553500e+04  6.553500e+04  1.700000e+01   1.200000e+08   

       Tot Fwd Pkts  Tot Bwd Pkts  TotLen Fwd Pkts  TotLen Bwd Pkts  \
count  7.616509e+06  7.616509e+06     7.616509e+06     7.616509e+06   
mean   1.314120e+01  6.606282e+00     6.304985e+02     5.230745e+03   
std    1.009401e+03  3.239508e+02     3.153876e+04     6.971802e+05   
min    0.000000e+00  0.000000e+00     0.000000e+00     0.000000e+00   
25%    1.000000e+00  1.000000e+00     2.900000e+01     0.000000e+00   
50%    2.000000e+00  3.000000e+00     4.600000e+01     2.470000e+02   
75%    8.000000e+00  7.000000e+00     9.350000e+02     1.324000e+03   
max    2.594430e+05  2.919230e+05     8.302176e+06     6.554527e+08   

       Fwd Pkt Len Max  Fwd Pkt Len Min  Fwd Pkt Len Mean  Fwd Pkt Len Std  \
count     7.616509e+06     7.616509e+06      7.616509e+06     7.616509e+06   
mean      2.196190e+02     1.186438e+01      5.794405e+01     8.231711e+01   
std       3.273044e+02     2.411128e+01      7.212265e+01     1.384465e+02   
min       0.000000e+00     0.000000e+00      0.000000e+00     0.000000e+00   
25%       2.800000e+01     0.000000e+00      1.033333e+01     0.000000e+00   
50%       4.200000e+01     0.000000e+00      3.900000e+01     1.154701e+01   
75%       9.350000e+02     3.500000e+01      2.337500e+02     4.675000e+02   
max       2.336000e+04     1.472000e+03      4.660441e+03     7.137621e+03   

       Bwd Pkt Len Max  Bwd Pkt Len Min  Bwd Pkt Len Mean  Bwd Pkt Len Std  \
count     7.616509e+06     7.616509e+06      7.616509e+06     7.616509e+06   
mean      3.571852e+02     2.843344e+01      1.150348e+02     1.310117e+02   
std       5.177774e+02     5.191029e+01      1.684583e+02     2.062547e+02   
min       0.000000e+00     0.000000e+00      0.000000e+00     0.000000e+00   
25%       0.000000e+00     0.000000e+00      0.000000e+00     0.000000e+00   
50%       2.470000e+02     0.000000e+00      7.700000e+01     1.299214e+02   
75%       7.880000e+02     6.200000e+01      1.781667e+02     2.595107e+02   
max       6.516000e+04     3.684000e+03      3.037508e+04     2.244841e+04   

        Flow Byts/s   Flow Pkts/s  Flow IAT Mean  Flow IAT Std  Flow IAT Max  \
count  7.586821e+06  7.616509e+06   7.616509e+06  7.616509e+06  7.616509e+06   
mean            inf           inf   2.225967e+06  1.134660e+06  5.318438e+06   
std             inf           inf   1.044042e+07  3.904521e+06  1.529950e+07   
min    0.000000e+00 -2.000000e+06  -1.000000e+00  0.000000e+00 -1.000000e+00   
25%    1.475935e+01  5.998456e+00   2.245000e+03  0.000000e+00  2.829000e+03   
50%    1.375624e+03  1.381478e+02   2.622900e+04  3.213559e+02  3.661800e+04   
75%    1.133630e+05  6.734007e+03   5.980003e+05  1.458536e+06  4.173576e+06   
max             inf           inf   1.199993e+08  8.480808e+07  1.199997e+08   

       Flow IAT Min   Fwd IAT Tot  Fwd IAT Mean   Fwd IAT Std   Fwd IAT Max  \
count  7.616509e+06  7.616509e+06  7.616509e+06  7.616509e+06  7.616509e+06   
mean   1.728327e+06  1.107273e+07  2.488425e+06  1.164599e+06  4.816895e+06   
std    1.034646e+07  3.048820e+07  1.061671e+07  4.396200e+06  1.493230e+07   
min   -1.300000e+01  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    6.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
50%    5.300000e+01  1.410000e+03  5.044167e+02

In [13]:
unique_list = []
n = df.columns.shape[0]
for i in range(n):
    column = df.columns[i]
    unique_list.append(df[column].nunique().compute())
    print(' '*100, end='\r')
    print(f'{column:20s} {unique_list[i]}')
    print('[' + (i+1) * '=' + '>' + (n-i-1) * ' ' + ']' + '\t\t' + f'{i+1}/{n}', end='\r')

Unnamed: 0           6553320                                                                        
Flow ID              4715999                                                                        
Src IP               36760                                                                          
Src Port             64849                                                                          
Dst IP               34321                                                                          
Dst Port             59091                                                                          
Protocol             3                                                                              
Timestamp            77598                                                                          
Flow Duration        2722954                                                                        
Tot Fwd Pkts         2501                                                                  

Unnamed: 0 	 6553320

Flow ID 	 4715999

Src IP 	 36760

Src Port 	 64849

Dst IP 	 34321

Dst Port 	 59091

Protocol 	 3

Timestamp 	 77598

Flow Duration 	 2722954

Tot Fwd Pkts 	 2501

Tot Bwd Pkts 	 2854

TotLen Fwd Pkts 	 14820

TotLen Bwd Pkts 	 60048

Fwd Pkt Len Max 	 3431

Fwd Pkt Len Min 	 479

Fwd Pkt Len Mean 	 98698

Fwd Pkt Len Std 	 210758

Bwd Pkt Len Max 	 3787

Bwd Pkt Len Min 	 444

Bwd Pkt Len Mean 	 165786

Bwd Pkt Len Std 	 243233

Flow Byts/s 	 3569113

Flow Pkts/s 	 3126852

Flow IAT Mean 	 3092172

Flow IAT Std 	 3619380

Flow IAT Max 	 1783978

Flow IAT Min 	 344330

Fwd IAT Tot 	 2260236

Fwd IAT Mean 	 2531071

Fwd IAT Std 	 2801288

Fwd IAT Max 	 1526963

Fwd IAT Min 	 443349

Bwd IAT Tot 	 2224031

Bwd IAT Mean 	 2444275

Bwd IAT Std 	 2605434

Bwd IAT Max 	 1426977

Bwd IAT Min 	 374503

Fwd PSH Flags 	 2

Bwd PSH Flags 	 2

Fwd URG Flags 	 1

Bwd URG Flags 	 1

Fwd Header Len 	 4225

Bwd Header Len 	 5132

Fwd Pkts/s 	 3013865

Bwd Pkts/s 	 2722380

Pkt Len Min 	 175

Pkt Len Max 	 4247

Pkt Len Mean 	 261546

Pkt Len Std 	 372908

Pkt Len Var 	 379741

FIN Flag Cnt 	 2

SYN Flag Cnt 	 2

RST Flag Cnt 	 2

PSH Flag Cnt 	 2

ACK Flag Cnt 	 2

URG Flag Cnt 	 2

CWE Flag Count 	 2

ECE Flag Cnt 	 2

Down/Up Ratio 	 213

Pkt Size Avg 	 204655

Fwd Seg Size Avg 	 71745

Bwd Seg Size Avg 	 137298

Fwd Byts/b Avg 	 1

Fwd Pkts/b Avg 	 1

Fwd Blk Rate Avg 	 1

Bwd Byts/b Avg 	 1

Bwd Pkts/b Avg 	 1

Bwd Blk Rate Avg 	 1

Subflow Fwd Pkts 	 2501

Subflow Fwd Byts 	 14820

Subflow Bwd Pkts 	 2854

Subflow Bwd Byts 	 60048

Init Fwd Win Byts 	 9520

Init Bwd Win Byts 	 13109

Fwd Act Data Pkts 	 749

Fwd Seg Size Min 	 10

Active Mean 	 704693

Active Std 	 542822

Active Max 	 580373

Active Min 	 287446

Idle Mean 	 863197

Idle Std 	 564679

Idle Max 	 599357

Idle Min 	 678358

Label 	 2

In [14]:
# Amount of Null values 'Flow Byts/s'
df['Flow Byts/s'].isnull().sum().compute()

29688

In [17]:
# Amount of infinite values per label
df[ (df['Flow Byts/s'] == np.inf) & (df['Flow Pkts/s'] == np.inf)].groupby('Label').agg(['count']).compute()

Unnamed: 0 Flow ID Src IP Src Port Dst IP Dst Port Protocol Timestamp  \
            count   count  count    count  count    count    count     count   
Label                                                                          
Benign      18067   18067  18067    18067  18067    18067    18067     18067   

       Flow Duration Tot Fwd Pkts Tot Bwd Pkts TotLen Fwd Pkts  \
               count        count        count           count   
Label                                                            
Benign         18067        18067        18067           18067   

       TotLen Bwd Pkts Fwd Pkt Len Max Fwd Pkt Len Min Fwd Pkt Len Mean  \
                 count           count           count            count   
Label                                                                     
Benign           18067           18067           18067            18067   

       Fwd Pkt Len Std Bwd Pkt Len Max Bwd Pkt Len Min Bwd Pkt Len Mean  \
                 count           count           count            count   
Label                                                                     
Benign           18067           18067           18067            18067   

       Bwd Pkt Len Std Flow Byts/s Flow Pkts/s Flow IAT Mean Flow IAT Std  \
                 count       count       count         count        count   
Label                                                                       
Benign           18067       18067       18067         18067        18067   

       Flow IAT Max Flow IAT Min Fwd IAT Tot Fwd IAT Mean Fwd IAT Std  \
              count        count       count        count       count   
Label                                                                   
Benign        18067        18067       18067        18067       18067   

       Fwd IAT Max Fwd IAT Min Bwd IAT Tot Bwd IAT Mean Bwd IAT Std  \
             count       count       count        count       count   
Label                                                                 
Benign       18067       18067       18067        18067       18067   

       Bwd IAT Max Bwd IAT Min Fwd PSH Flags Bwd PSH Flags Fwd URG Flags  \
             count       count         count         count         count   
Label                                                                      
Benign       18067       18067         18067         18067         18067   

       Bwd URG Flags Fwd Header Len Bwd Header Len Fwd Pkts/s Bwd Pkts/s  \
               count          count          count      count      count   
Label                                                                      
Benign         18067          18067          18067      18067      18067   

       Pkt Len Min Pkt Len Max Pkt Len Mean Pkt Len Std Pkt Len Var  \
             count       count        count       count       count   
Label                                                                 
Benign       18067       18067        18067       18067       18067   

       FIN Flag Cnt SYN Flag Cnt RST Flag Cnt PSH Flag Cnt ACK Flag Cnt  \
              count        count        count        count        count   
Label                                                                     
Benign        18067        18067        18067        18067        18067   

       URG Flag Cnt CWE Flag Count ECE Flag Cnt Down/Up Ratio Pkt Size Avg  \
              count          count        count         count        count   
Label                                                                        
Benign        18067          18067        18067         18067        18067   

       Fwd Seg Size Avg Bwd Seg Size Avg Fwd Byts/b Avg Fwd Pkts/b Avg  \
                  count            count          count          count   
Label                                                                    
Benign            18067            18067          18067          18067   

       Fwd Blk Rate Avg Bwd Byts/b Avg Bwd Pkts/b Avg Bwd Blk Rate Avg  \
                  count          count          coun

---
## Insights Gained

### Not Informative Data
To gain further insights about the columns, we have determined the details and amount of unique values for column.
The following columns contain only one unique value:
- Fwd URG Flags 
- Bwd URG Flags 
- Fwd Byts/b Avg
- Fwd Pkts/b Avg
- Bwd Byts/b Avg
- Bwd Pkts/b Avg
- Fwd Blk Rate Avg
- Bwd Blk Rate Avg

Since these attributes can only take on one value, we cannot gain any information about the distinction between benign and ddos. Therefore, these columns will be removed. That means that we drop about 10% of the origin columns.

### Null-Values

The columns 'Flow Byts/s' is the only columns containing 29688 null values (NaN - Not a Number), which we discovered through the use of .info. There are to possible options for handling this:
- Option 1: Drop
  - Dropping the values seems sensible since we do not make any wrong assumptions about the data. Furthermore, the columns only contain 0.4% of the overall data.
- Option 2: Replace
  - Replacing the values does not lead to artifical changes within the timeline (timestamps) which might cause troubles for some models.

For now, we decided to use option 1.

## Infinity-Values
During the exploration of the data, we have noticed infinity values within 0.2% of the data of the columns 'Flow Byts/s' and 'Flow Pkts/s'. All of the rows with these values are labled 'benign'. For this reason, we decided to drop the rows as well.

---
## Cleaning

In [ ]:
df_inf = df[~da.isinf(df['Flow Byts/s'])]
df_nonull = df_inf.dropna(subset=['Flow Byts/s'])
df_clean = df_nonull.drop(['Fwd URG Flags', 'Bwd URG Flags', 'Fwd Byts/b Avg', 'Fwd Pkts/b Avg', 'Bwd Byts/b Avg', 'Bwd Pkts/b Avg', 'Fwd Blk Rate Avg', 'Bwd Blk Rate Avg'], axis=1)
df_clean.to_csv('cleaned_ds.csv', single_file = True, index = False)